identity-account/src/error.rs

```diff
@@ -76,6 +76,12 @@ pub enum Error {
   InvalidPrivateKey(String),
   #[error("method missing fragment")]
   MethodMissingFragment,
+  #[cfg(feature = "wasm")]
+  #[error("JsValue serialization error: {0}")]
+  SerializationError(String),
+  #[cfg(feature = "wasm")]
+  #[error("javascript function threw an exception: {0}")]
+  JsError(String),
```

Comment:
Regarding errors for the `Storage` interface, I think it's a bit unwieldy to restrict implementers to the top-level `AccountError`.

Maybe we can have a sub-error enum for the `Storage` and map them to a single `AccountError` internally? This would be similar to the `Error` in the diff crate, where each variant maps to a function.

```rust
#[derive(Debug, thiserror::Error, strum::IntoStaticStr)]
pub enum Error {
  #[error("Diff Error: {0}")]
  DiffError(String),
  #[error("Merge Error: {0}")]
  MergeError(String),
  #[error("Conversion Error: {0}")]
  ConversionError(String),
}
```

We could use strings too or retain the error `source` with e.g. `Box<dyn Error>` to preserve reporting, as mentioned in #636.

This would basically eliminate the need for these feature-gated `JsError`/`SerializationError` variants since we could define errors in the Wasm bindings themselves that are bubbled up to the correct variant with no loss of context (hopefully).

This would benefit future bindings and `Storage` implementers using the core Rust library too.

cc @olivereanderson

Reply:
There definitely needs to be a separate `StorageError`, and I really like your suggestion with `Box<dyn Error>`. Still, I think this needs a good amount of thought put into it to provide a reasonably abstract error that can accommodate not just stronghold and memory stores, which is why I would suggest doing this in a separate PR and leave the current approach as an acceptable temporary solution. I'd be happy to open an issue for it.

Reply:
Sure, it can be added to #669 or its own issue. I wouldn't merge this without having a tracking issue for what's left, however.